src/server/tweetFinder.js

```diff
@@ -73,30 +73,86 @@ const resultsToPresentableOutput = (results) => (
   }
 );
 
-const searchDatabase = (query, alreadyAttemptedRefresh = false) => {
-  const tweetSelection = 'SELECT *, in(\'TWEETED\').id AS authorId, in(\'TWEETED\').name AS authorName, in(\'TWEETED\').handle AS authorHandle FROM tweet WHERE content LUCENE :query ORDER BY date DESC UNWIND authorId, authorName, authorHandle LIMIT :limit';
-
-  return newPromiseChain()
-    .then(() => db.query(tweetSelection, { params: { query: `${query}~`, limit: 300 } }))
-    .then((tweetRecords) => refreshFromTwitterOrMakeTweets(alreadyAttemptedRefresh, query, tweetRecords))
+const searchDatabase = (searchObject, alreadyAttemptedRefresh = false) => (
+  newPromiseChain()
+    .then(() => Promise.all(
+      searchObject.paramTypes
+        .filter((paramType) => paramType.selected)
+        .map((paramType) => searchByParamType(searchObject, paramType.name))
+      )
+    )
+    .then((searchResults) => searchResults.reduce((previous, current) => previous.concat(current)))
+    .then((tweetRecords) => refreshFromTwitterOrMakeTweets(alreadyAttemptedRefresh, searchObject, tweetRecords))
     .then(
       (resolved) => resolved,
       (rejection) => console.warn('Major error querying the database.', rejection)
-    );
+    )
+);
+
+const searchByParamType = (searchObject, paramType) => {
+  switch (paramType) {
+    case 'keyword': return searchByKeyword(searchObject.query);
+    case 'author': return searchByAuthor(searchObject.query);
+    case 'mention': return searchByMention(searchObject.query);
+    case 'hashtag': return searchByHashtag(searchObject.query);
+  }
+
+  throw(`Invalid paramType for database Tweet searching: '${paramType}'. Should be [author, hashtag, keyword, mention].`);
+};
+
+const searchByKeyword = (keyword) => {
+  const tweetSelection = makeTweetQuerySelectingFrom(
+    'SELECT FROM tweet WHERE content LUCENE :query'
+  );
+  return db.query(tweetSelection, { params: { query: `${keyword}~`, limit: 300 } });
 };
 
-const refreshFromTwitterOrMakeTweets = (alreadyAttemptedRefresh, query, tweetRecords) => {
+const searchByAuthor = (keyword) => {
+  const tweetSelection = makeTweetQuerySelectingFrom(
+    'TRAVERSE out(\'TWEETED\') FROM (SELECT FROM Tweeter WHERE name LUCENE :query OR handle LUCENE :query)'
+  );
+  return db.query(tweetSelection, { params: { query: `${keyword}~`, limit: 300 } });
+};
+
+const searchByMention = (keyword) => {
+  const tweetSelection = makeTweetQuerySelectingFrom(
+    'TRAVERSE in(\'MENTIONS\') FROM (SELECT FROM Tweeter WHERE name LUCENE :query OR handle LUCENE :query)'
+  );
+  return db.query(tweetSelection, { params: { query: `${keyword}~`, limit: 300 } });
+};
+
+const searchByHashtag = (keyword) => {
+  const tweetSelection = makeTweetQuerySelectingFrom(
+    'TRAVERSE in(\'HAS_HASHTAG\') FROM (SELECT FROM hashtag WHERE content LUCENE :query)'
+  );
+  return db.query(tweetSelection, { params: { query: `${keyword}~`, limit: 300 } });
+};
+
+const makeTweetQuerySelectingFrom = (from) => (

```

Comment:
Not according to any guides I've seen, nor is it highlighted in WebStorm. I think it's safe, but can change it if we're not sure?